Update docs for Key Packages (#2049)
openmls/openmls: book/src/user_manual/create_key_package.md

# Key Packages

> The MLS protocol is designed in a way that allows members to be added to groups without the said members being online (because more generally: MLS communication is entirely designed to be asynchronous).

Reviewer: ```suggestion
The MLS protocol is designed to be asynchronous: It allows members to be added to groups without said members being online.
```

The general flow is like this:

1. **Client A** generates key packages and registers them with the Delivery Service
2. **Client B** decides to add **Client A** to a group, so it contacts the Delivery Service and downloads one of **Client A**'s key packages
3. The delivery service marks the `KeyPackage` as consumed (more on this below)
4. **Client B** uses the **KeyPackage** to generate a commit that adds **Client A** to the group

> Each key package has corresponding private keys that are meant to be stored locally only. Therefore OpenMLS defines a struct called `KeyPackageBundle`, which consists of:

Reviewer: ```suggestion
Each key package has corresponding private keys that are stored locally in a `KeyPackageBundle`, which consists of:
```

> - `key_package` (`KeyPackage`): The public material that should be registered at the Delivery Service

Reviewer: ```suggestion
- `key_package` (`KeyPackage`): Public key material and information required to add the client to a group
```

Author: I have a question: Is omitting that this is the bit to be uploaded to the DS intentional? Or in other words what's the expectation of the reader's background here?

Cause I wrote in a way that's easy to understand for someone that never read the RFC, but I also know I might be making too many assumptions with these statements :v

> - `private_init_key` (`HpkePrivateKey`): Private stuff that should never leave the local database

Reviewer: ```suggestion
- `private_init_key` (`HpkePrivateKey`): The private key required to decrypt the `Welcome` message that adds the client to a group
```

> - `private_encryption_key` (`EncryptionPrivateKey`): Private stuff that should never leave the local database

Reviewer: ```suggestion
- `private_encryption_key` (`EncryptionPrivateKey`): The private key used for group operations after the client was added
```

Key packages are meant to be used only once ([unless it's a last resort KeyPackage, if you implement those](https://www.rfc-editor.org/rfc/rfc9420.html#name-keypackage-reuse)), which means a few things:

1. The delivery service should make sure a key package can only be downloaded once (unless it's a last resort one)
2. Each client should maintain many key packages registered at the Delivery Service so the client can be added to many groups while it's offline (without using the last resort key package, if it exists)

> 3. Each client should continuously replenish the list of key packages registered at the Delivery Service, replacing any consumed / nearing expiry / expired ones

Reviewer: ```suggestion
Key packages are meant to be used only once ([unless it's a last resort KeyPackage](https://www.rfc-editor.org/rfc/rfc9420.html#name-keypackage-reuse)). The Delivery Service should ensure that each (non-last resort) KeyPackage is only downloaded once.

Clients should register multiple KeyPackages and one or more last-resort KeyPackages with the DS, and replenish them sufficiently frequently. This is to ensure that the DS's KeyPackage supply is not exhausted (either through use or KeyPackage expiry) and that key material stays fresh. 
```

Each key package has a lifetime attached to it ([default is `3 * 28` days](../../../openmls/src/key_packages/lifetime.rs#L13)), so the client should keep that in mind when it comes to replenishing registered KeyPackages at the Delivery Service.

Finally, a key package is identified within the protocol by a ["hash reference"](https://www.rfc-editor.org/rfc/rfc9420.html#name-hash-based-identifiers).

## Creating `KeyPackageBundle`s

You need to choose a few parameters when creating `KeyPackageBundle`s:

- `ciphersuites: &[CiphersuiteName]`: A list of ciphersuites supported by the client.
- `extensions: Vec<Extensions>`: A list of supported extensions.

> You must specify at least one ciphersuite and not advertise ciphersuites they do not support.

Reviewer: ```suggestion
The client must specify at least one ciphersuite per KeyPackage and not advertise ciphersuites the client does not support.
```

You should specify all extensions they support. See the documentation of extensions for more details.

Here's how you can generate a `KeyPackageBundle`:

```rust
// The following needs to be pre-defined:
// - CIPHERSUITE (`Ciphersuite`)
// - provider (`OpenMlsProvider`)
// - signature_key_pair (`SignatureKeyPair`)
// - credential (`BasicCredential`)

let key_package_bundle = KeyPackage::builder()
	.build(
		CIPHERSUITE,
		&provider, 
		&signature_key_pair,
		CredentialWithKey {
			credential: credential.clone().into(),
			signature_key: signature_key_pair.public().into(),
		},
	)?;

// This is the key package's identifier (upload this to the Delivery Service))

> let hash_ref: Vec<u8> = key_package_bundle.key_package().hash_ref(provider.crypto())?.tls_serialize_detached()?

Reviewer: No need to upload this. The DS can just create the hash itself.

// This is the exact data another client would need to add this client to a group (upload this to the Delivery Service)
let key_package: Vec<u8> = key_package_bundle.key_package().tls_serialize_detached()?;

// This is the expiry date of the key pacakge
let expires_at: u64 = key_package_bundle.key_package().life_time().not_after();

> ```

Reviewer: This should be safe to remove. It's already mentioned above that KeyPackages have lifetimes and `life_time`, as well as `not_after` are self-explanatory.

The private parts of the key package are automatically stored in the database.

## Getting existing key packages

OpenMLS does not provide an API for this; While you're implementing `StorageProvider` for your local database, you can implement a `get_key_packages()` function that queries the storage to get that for you.

## Deleting a key package (identified by hash reference)

```rust
let hash_ref: Vec<u8> = ...;

provider.storage().delete_key_package(
	&KeyPackageRef::tls_deserialize_exact_bytes(hash_ref.as_slice())?
)?;
```

## Appendix: `KeyPackage` contents

A `KeyPackage` (the public part of `KeyPackageBundle`) consists of:

- A public HPKE encryption key to enable MLS' basic group key distribution feature
- The lifetime throughout which the key package is valid
- Information about the client's capabilities (i.e., which features of MLS it supports)
- Any extension that the client wants to include

> - One of the client's [credentials](./identity.md), as well as a signature over the whole key package using the private key corresponding to the credential's signature public key

Reviewer: It should be enough to provide a link to the code when you're listing the content of `KeyPackageBundle`.
